api/models/DeviceHistory.ts

```diff
@@ -113,5 +121,93 @@ export default function (
     });
   };
 
+  DeviceHistory.updateDeviceSettings = async function (
+    deviceId: DeviceId,
+    groupId: GroupId,
+    newSettings: ApiDeviceHistorySettings,
+    setBy: DeviceHistorySetBy
+  ): Promise<ApiDeviceHistorySettings> {
+    const currentSettingsEntry: DeviceHistory = await this.findOne({
+      where: {
+        DeviceId: deviceId,
+        GroupId: groupId,
+        location: { [Op.ne]: null },
+      },
+      order: [["fromDateTime", "DESC"]],
+    });
+
+    const currentSettings: ApiDeviceHistorySettings =
+      currentSettingsEntry?.settings ?? ({} as ApiDeviceHistorySettings);
+    const { settings, changed } = mergeSettings(
+      currentSettings,
+      newSettings,
+      setBy
+    );
+
+    const synced = setBy === "automatic";
+    // add to device history ledger
+    if (changed) {
+      await this.create({
+        ...currentSettingsEntry.get({ plain: true }),
+        DeviceId: deviceId,
+        GroupId: groupId,
+        fromDateTime: new Date(),
+        setBy,
+        settings,
+      });
+    } else if (synced) {
+      // in place only if the device already had the settings so no change
+      await currentSettingsEntry.update("settings", settings);
+    }
+
+    return settings;
+  };
+
   return DeviceHistory;
 }
+// Function to merge settings using "Last Write Wins"
+function mergeSettings(
+  currentSettings: ApiDeviceHistorySettings,
+  incomingSettings: ApiDeviceHistorySettings,
+  setBy: DeviceHistorySetBy
+): { settings: ApiDeviceHistorySettings; changed: boolean } {
+  const mergedSettings: ApiDeviceHistorySettings = { ...currentSettings };
+
+  let changed = false;
+  for (const key in incomingSettings) {
+    if (incomingSettings.hasOwnProperty(key)) {
```

Comment:
Small nit: when is this ever not true, given that we're iterating keys in `incomingSettings`.  In the case where you're worried about inherited keys from up the prototype chain you can use `Object.keys` which doesn't iterate inherited keys, only ones that are `hasOwnProperty`.